azurelinuxagent/ga/exthandlers.py

```diff
@@ -276,6 +288,78 @@ def run(self):
+        if fabric_goal_state_changed:
+            self.protocol.set_fast_track(False)
+
+        # Fabric goal state takes precedence over VMArtifactsProfile blob goal state
+        # So, process the Fabric goal state first
+        self.handle_ext_handlers(fabric_goal_state_changed)
+        self.last_incarnation = incarnation
+
+        # Now process the VMArtifactsProfile if it changed. Note that just because the VMArtifactsProfile changed
+        # doesn't mean the FastTrack goal state changed
+        if fast_track_changed and self.artifacts_profile is not None:
+            vm_artifacts_ext_conf = self.artifacts_profile.transform_to_extensions_config()
+            self.process_vm_artifacts_profile_for_fast_track(sequence_number, vm_artifacts_ext_conf)
+        elif self.protocol.client.get_use_fast_track():
+            # Our last goal state was via VMArtifactsProfile, so reset the extension handlers
+            self.ext_conf = self.protocol.client.get_ext_conf(read_always=True)
+            self.ext_handlers = self.ext_conf.ext_handlers
+
+    def determine_fabric_changed(self, incarnation):
+        fabric_goal_state_changed = False
+
+        if self.last_incarnation != incarnation:
+            logger.info('Goal state changed with incarnation={0}(last:{1})'.
+                        format(str(incarnation), str(self.last_incarnation)))
+            fabric_goal_state_changed = True
+
+        return fabric_goal_state_changed
+
+    def determine_fast_track_changed(self, sequence_number):
+        fast_track_changed = False
+
+        if sequence_number is not None and self.last_vm_artifacts_seqno != sequence_number and \
+                conf.get_extensions_fast_track_enabled():
+            logger.info('VMArtifactsProfile changed with sequence number={0}(last:{1})'.
+                        format(str(sequence_number), str(self.last_vm_artifacts_seqno)))
+            fast_track_changed = True
+        return fast_track_changed
+
+    def process_vm_artifacts_profile_for_fast_track(self, sequence_number, vm_artifacts_ext_conf):
+        if vm_artifacts_ext_conf is not None:
+            # We need to add the status upload blob from Fabric goal state because we don't receive this
+            # from the VMArtifactsProfile blob
+            vm_artifacts_ext_conf.status_upload_blob = self.ext_conf.status_upload_blob
+            vm_artifacts_ext_conf.status_upload_blob_type = self.ext_conf.status_upload_blob_type
+            if vm_artifacts_ext_conf.ext_handlers is not None and \
+                    len(vm_artifacts_ext_conf.ext_handlers.extHandlers) > 0:
+                logger.info('FastTrack goal state changed.')
+                self.protocol.set_fast_track(True, sequence_number)
+
+                # Reset the extension config and reload it, which will load it with the FastTrack extensions
+                self.ext_conf = self.protocol.client.get_ext_conf(read_always=True)
```

Comment:
Please correct me if I'm wrong, but all this function `self.protocol.client.get_ext_conf(read_always=True)` is doing is reading again from the ExtensionConfig file and setting the `self.protocol.client.ext_conf` to the new ExtensionConfig. But isin't the `vm_artifacts_ext_conf` same thing? Why are you reading the same config again from the file rather than just using the same object `vm_artifacts_ext_conf`?